src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs

```diff
@@ -12,202 +16,340 @@ namespace System.Buffers
+        private static SequencePosition GetBufferCrossSegment(int startIndex, object endObject, ref ReadOnlySequenceSegment<T> startSegment, ref int length)
         {
-            IMemoryList<T> nextSegment = startMemoryList.Next;
-            int currentLength = startMemoryList.Memory.Length - startIndex;
+            Debug.Assert(startSegment != null);
+
+            ReadOnlySequenceSegment<T> nextSegment = startSegment.Next;
+            int currentLength = startSegment.Memory.Length - startIndex;
 
-            while (currentLength == 0 && nextSegment != null)
+            while (currentLength == 0 && nextSegment != endObject && nextSegment != null)
             {
                 // Skip empty Segments
-                startMemoryList = nextSegment;
+                startSegment = nextSegment;
                 nextSegment = nextSegment.Next;
-                currentLength = startMemoryList.Memory.Length;
+                currentLength = startSegment.Memory.Length;
             }
 
             length = currentLength;
-            SequencePosition next;
-            if (nextSegment == null)
+            while (nextSegment != null && nextSegment.Memory.Length == 0)
             {
-                if (end.GetObject() != null)
-                    ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();
+                // Skip empty Next Segments
+                if (nextSegment == endObject)
+                {
+                    // Validated end reached, clear next as empty
+                    endObject = nextSegment = null;
+                    break;
+                }
 
-                next = default;
+                nextSegment = nextSegment.Next;
             }
-            else
+
+            if (nextSegment != null)
             {
-                next = new SequencePosition(nextSegment, 0);
+                return new SequencePosition(nextSegment, 0);
             }
+            else
+            {
+                if (endObject != null)
+                    ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();
 
-            return next;
+                return default;
+            }
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        internal SequencePosition Seek(in SequencePosition start, in SequencePosition end, int count, bool checkEndReachable = true)
+        internal SequencePosition Seek(in SequencePosition start, in SequencePosition end, int count)
         {
             GetTypeAndIndices(start.GetInteger(), end.GetInteger(), out SequenceType type, out int startIndex, out int endIndex);
 
-            object startSegment = start.GetObject();
+            object startObject = start.GetObject();
             bool notInRange = endIndex - startIndex < count;
-            if (type == SequenceType.IMemoryList)
+            if (type == SequenceType.MultiSegment)
```

Comment:
Add a comment here why other segment types just work. And in the other overload too.

Reply:
Can un-nest these loops one level; will make it cleared

Reply:
Unnested `if` and added more comments